File: ios/flutter_compress/Sources/flutter_compress/Thumbnailer.swift
import AVFoundation
import UIKit

/// Extracts a single JPEG thumbnail frame from a video.
enum Thumbnailer {
  static func generate(
    dir: URL, path: String, positionMs: Int64, quality: Int, maxWidth: Int?
  ) throws -> String {
    let generator = AVAssetImageGenerator(asset: AVURLAsset(url: URL(fileURLWithPath: path)))
    generator.appliesPreferredTrackTransform = true
    if let maxWidth = maxWidth {
      generator.maximumSize = CGSize(width: maxWidth, height: 0)
    }
    let cg = try generator.copyCGImage(
      at: CMTime(value: positionMs, timescale: 1000), actualTime: nil)
    guard let data = UIImage(cgImage: cg).jpegData(compressionQuality: CGFloat(quality) / 100.0)
    else {
      throw NSError(
        domain: "flutter_compress", code: 2,
        userInfo: [NSLocalizedDescriptionKey: "JPEG encode failed"])
    }
    let out = dir.appendingPathComponent("thumb_\(Int(Date().timeIntervalSince1970 * 1000)).jpg")
    try data.write(to: out)
    return out.path
  }
}

File: android/src/main/kotlin/com/compress/all/flutter_compress/Thumbnailer.kt
package com.compress.all.flutter_compress

import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import java.io.File
import java.io.FileOutputStream

/** Extracts a single JPEG thumbnail frame from a video. */
object Thumbnailer {
    fun generate(
        dir: File,
        path: String,
        positionMs: Long,
        quality: Int,
        maxWidth: Int?,
    ): String {
        val r = MediaMetadataRetriever()
        try {
            r.setDataSource(path)
            var bmp = r.getFrameAtTime(positionMs * 1000)
                ?: error("Could not extract frame")
            if (maxWidth != null && bmp.width > maxWidth) {
                val h = (bmp.height * maxWidth.toFloat() / bmp.width).toInt()
                bmp = Bitmap.createScaledBitmap(bmp, maxWidth, h, true)
            }
            val out = File(dir, "thumb_${System.nanoTime()}.jpg")
            FileOutputStream(out).use { bmp.compress(Bitmap.CompressFormat.JPEG, quality, it) }
            return out.absolutePath
        } finally {
            r.release()
        }
    }
}
